# Deskriptive mål & grupperinger

## Deskriptive mål

Én måde at udforske et datasæt er ved at se nærmere på forskellige deskriptive mål for variable i datasættet.

Metoden `.describe()` danner deskriptive mål for alle (kompatible) variable i datasættet:

In [1]:
import pandas as pd

ess2018 = pd.read_csv("https://github.com/CALDISS-AAU/course_ddf/raw/master/datasets/ESS2018DK_subset.csv")

In [2]:
ess2018.describe()

idno      netustm      ppltrst        yrbrn       eduyrs  \
count    1285.000000  1134.000000  1282.000000  1285.000000  1278.000000   
mean   110979.542412   227.431217     7.079563  1967.047471    13.996088   
std     25534.338367   193.406614     1.800747    18.294410     5.746878   
min      5816.000000     0.000000     0.000000  1929.000000     0.000000   
25%     93707.000000    90.000000     6.000000  1952.000000    11.000000   
50%    112877.000000   150.000000     7.000000  1966.000000    14.000000   
75%    131072.000000   300.000000     8.000000  1982.000000    17.000000   
max    150446.000000  1020.000000    10.000000  2003.000000    40.000000   

             wkhct       wkhtot       grspnum        frlgrsp        inwtm  
count  1217.000000  1258.000000  7.150000e+02     252.000000  1280.000000  
mean     33.806902    36.441176  4.368549e+04   37255.230159    63.323437  
std       9.399469    11.762775  1.221933e+05   33449.267184    26.102194  
min       0.000000     0.000000  0.000000e+00       0.000000    18.000000  
25%      35.000000    35.000000  2.600000e+04   28375.000000    51.000000  
50%      37.000000    37.000000  3.200000e+04   35000.000000    59.000000  
75%      37.000000    41.000000  4.100000e+04   40000.000000    70.000000  
max      80.000000    90.000000  3.000000e+06  500000.000000   613.000000

Metoden fungerer også på enkeltvariable:

In [21]:
ess2018['wkhtot'].describe()

count    1502.000000
mean       36.171771
std        13.205436
min         0.000000
25%        34.000000
50%        37.000000
75%        42.000000
max       168.000000
Name: wkhtot, dtype: float64

`.describe()` danner følgende mål:
- `count`: Antal svar ekskl. missing
- `mean`: Middelværdien
- `std`: Standardafvigelsen
- `min`: Minimumværdien
- `25%`: 1. kvartil
- `50%`: 2. kvartil
- `75%`: 3. kvartil
- `max`: Maksimumværdien

Der knytter sig også en række metoder til at danne de enkelte deskriptive mål.

Herunder printes fx middelværdien, standardafvigelsen, minimums- og maksimumsværdien for variablen `weight`:

In [22]:
print(ess2018['wkhtot'].mean())
print(ess2018['wkhtot'].std())
print(ess2018['wkhtot'].min())
print(ess2018['wkhtot'].max())

36.17177097203729
13.20543573826518
0.0
168.0


Ovenstående værdier fortæller, at `wkhtot` indeholder værdier fra 0.0 til 168.0 med en middelværdi på 36.17 og en standardafvigelse på 13.21 (middelværdien af de afstande, som hver observation afviger fra middelværdien).

### Deskriptive mål og kategoriske variable

Sammenlignes outputtet fra `.describe()` med indholdet af datasættet ses, at der ikke udregnes deskriptive mål for alle variable. 

Når `.describe()` bruges på et hel dataframe, udregnes kun mål for numeriske variabeltyper; altså variable, der kun består af tal.

Metoden kan dog også bruges på en kategorisk variabel som `ppltrst`:

In [23]:
ess2018['vote'].describe()

count     1566
unique       3
top        Yes
freq      1332
Name: vote, dtype: object

## Grupperinger: Opsummeringer på gruppeniveau (groupby)

En effektiv måde at udforske og udføre indledende deskriptive analyser af sine data, er ved at undersøge de grupperinger, som data indeholder.

Her skal vi se på nogen forskellige måder, hvor der kan udregnes deskriptive mål og dannes visualiseringer for grupper i data

Vi starter med at indlæse de nødvendige pakker og data. Aldersvariablen (som også blev tilføjet i lektion 3) tilføjes igen:

### Brug af split-apply til at opsummere data

Metoden `.groupby()` grupperer datasæt efter de givne variable:

In [24]:
grouped_df = ess2018.groupby(['gndr'])

Selve objektet, som bliver dannet, indeholder ikke information, der bare kan kaldes frem direkte:

In [25]:
grouped_df

Dog kan vi danne de samme deskriptive mål, som vi kan for hele datasættet, men hvor de opsummeres på gruppeniveauet:

In [26]:
grouped_df.mean()

idno     netustm   ppltrst        yrbrn     eduyrs  \
gndr                                                                  
Female  111135.282369  227.564669  7.140496  1967.834711  13.847010   
Male    110552.542553  236.069645  6.996437  1968.598109  13.534524   

            wkhct     wkhtot        grspnum       frlgrsp      inwtm  
gndr                                                                  
Female  31.939850  33.561404   53165.458689  45699.366667  64.078621  
Male    34.324393  38.354523  128033.880597  84030.717742  61.468490

Det samme kan gøres for enkeltvariable:

In [27]:
grouped_df['wkhtot'].mean()

gndr
Female    33.561404
Male      38.354523
Name: wkhtot, dtype: float64

En lang række metoder kan bruges på grupperede data. Se dem alle her: https://pandas.pydata.org/pandas-docs/stable/user_guide/groupby.html.

- `mean()`: Middelværdi for grupperne

- `size()`: Størrelse af grupperne

- `count()`: Tællinger inden for grupperne

- `describe()`: Deskriptive mål inden for grupperne

- `min()`: Minimum for grupperne

- `max()`: Maximum for grupperne

Metoden `value_counts()` kan bruges til at lave optællinger af kategoriske inden for grupperne.

Herunder laves optælling for variablen `vote` fordelt på køn:

In [28]:
grouped_df['vote'].value_counts()

gndr    vote                
Female  Yes                     626
        Not eligible to vote     63
        No                       34
Male    Yes                     706
        Not eligible to vote     77
        No                       60
Name: vote, dtype: int64

# ØVELSE 3: Deskriptive mål

1. Brug `.describe()` eller passende metoder til at finde ud af følgende:

- Hvor gammel er den yngste person i datasættet?
- Hvor gammel er den ældste person i datasættet?
- Hvad er gennemsnitsalderen i datasættet?

- *Husk at datasættet er fra 2018.*


2. Brug `.groupby()` til at finde ud af, om kvinderne eller mændende i datasættet har flest års uddannelse i gennemsnit.
